src/Authoring/WinRT.SourceGenerator/AotOptimizer.cs

```diff
@@ -20,11 +20,15 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
         {
             var properties = context.AnalyzerConfigOptionsProvider.Select(static (provider, _) => (provider.IsCsWinRTAotOptimizerEnabled(), provider.IsCsWinRTComponent()));
 
-            var vtableAttributesToAdd = context.SyntaxProvider.CreateSyntaxProvider(
+            var vtableAttributesOrDiagnostics = context.SyntaxProvider.CreateSyntaxProvider(
                     static (n, _) => NeedVtableAttribute(n),
                     static (n, _) => GetVtableAttributeToAdd(n)
                 ).Where(vtableAttribute => vtableAttribute != null);
 
+            var diagnostics = vtableAttributesOrDiagnostics.Where(vtableAttribute => vtableAttribute.Diagnostic != null);
+            var vtableAttributesToAdd = vtableAttributesOrDiagnostics.Where(vtableAttribute => vtableAttribute.Diagnostic == null);
+
+            context.RegisterImplementationSourceOutput(diagnostics.Collect().Combine(properties), GenerateDiagnostics);

```

Comment:
Can we move this to a diagnostic analyzer? Emitting diagnostics from generators is highly frowned upon.

Reply:
Yea I was debating on this, I did it here initially because I wasn't sure if it was worth doing 2 different analysis for the same syntax for the same thing for one diagnostic.  But yea I can move it to a different analyzer.

Reply:
I'm hoping that in the analyzer we'd be able to do a very minimal amount of processing, as we only need to discover whether these types are exposed in some way, but not do all the actual work to actually compute the vtables and whatnot. We can also probably optimize the way we compare eg. types and whatnot, but resolving symbols from the compilation start and then comparing those, rather than formatting types and comparing their names. But also in general analyzers run out of process and don't impact the core IDE inner loop, so the Roslyn teams recommends using them for diagnostics whenever possible.

Reply:
Addressed.